### PR TITLE
Share the 200-char confirmation preview truncation across tools (#1292)

### LINKED ISSUE #1292: dry-violation: the 200-char confirmation preview is hand-rolled in four tool files
## What

Four separate tool implementations build their confirmation-dialog preview with the same hand-rolled truncation expression, character-for-character:

```ts
const preview = `${X.substring(0, 200)}${X.length > 200 ? '...' : ''}`;
```

It is a small expression, but it encodes three decisions that are supposed to be uniform across every tool confirmation the user sees — the cut length (200), the ellipsis form (`...` rather than `…`), and the "only append when actually truncated" rule. Right now those decisions live in four places with nothing tying them together, so changing the preview length or switching to a real ellipsis means finding all four by grep, and a fifth tool added tomorrow will copy whichever one its author happened to look at.

This is the same shared-leaf-helper shape the repo has already applied several times (#1227 collapsible toggles, #1214 tool→icon map, #1213 `buildCompactionEntry`, #1140 `toFileEntry`, #1139 `createDebouncedSave`), so the pattern and the destination convention are both established — what's missing is a decision about where this particular helper belongs.

## Evidence

Detection: `grep -rn "substring(0, 200)" src --include=*.ts`.

- `src/tools/vault/write-file-tool.ts:45` — `write_file` confirmation preview
- `src/tools/vault-tools-extended.ts:152` — `append_file` confirmation preview
- `src/tools/memory-tool.ts:34` — memory-write confirmation preview
- `src/tools/skill-tools.ts:174` — `create_skill` confirmation preview (truncates `params.description` rather than `params.content`, but the expression is otherwise identical)

A fifth site, `src/services/vault-analyzer.ts:369`, uses `substring(0, 200)` for a debug log snippet. That one is unrelated — it has no ellipsis and is not user-facing. It should **not** be folded into the helper.

## Proposed unit of work

- Add a `truncateForPreview(text: string, max = 200): string` helper returning `text.length > max ? text.slice(0, max) + '...' : text`.
- Put it in a leaf module both the tool files and any future caller can import without tripping `npm run lint:cycles`. `src/utils/format-utils.ts` is the natural home — it already holds `formatFileSize` and the timestamp formatters, and it imports nothing from `src/`. Confirm that choice rather than inheriting it; the alternative is a new `src/utils/text-utils.ts` if `format-utils` is meant to stay date/size-only.
- Replace the four call sites above with calls to the helper. Keep `skill-tools.ts` passing `params.description` — the argument differs, the logic doesn't.
- Add a unit test covering the three boundaries: under the limit (unchanged), exactly at the limit (no ellipsis), over the limit (cut plus ellipsis). The current behaviour appends nothing at exactly 200 chars; preserve that.
- Leave the visible output byte-identical. This is a de-duplication, not a chance to switch `...` to `…` — if a real ellipsis is wanted, that is a separate user-facing change with its own i18n implications.

## Out of scope

- `src/services/vault-analyzer.ts:369` — a debug-log snippet with no ellipsis and no user-facing surface.
- Changing the 200-character limit, the ellipsis character, or any confirmation wording.
- The surrounding `confirmationMessage` / `t()` calls in each tool; only the preview expression moves.
- The other duplicate blocks flagged in the same sweep — those are tracked separately.

---

_Filed by the `audit-architecture` skill. If this isn't worth doing, close as **Not planned** / `wontfix` — the skill checks closed-not-planned and won't refile._

### PR DESCRIPTION
**Pull request by allenhutchison**
Fixes #1292

Four tool files hand-rolled the identical confirmation-preview expression — encoding the cut length (200), the ASCII `...` ellipsis, and the "only append when actually truncated" rule separately in each:

- `write_file` → `src/tools/vault/write-file-tool.ts`
- `append_content` → `src/tools/vault/append-content-tool.ts`
- `update_memory` → `src/tools/memory-tool.ts`
- `create_skill` → `src/tools/skill-tools.ts` (truncates `params.description`, not content — argument differs, logic doesn't)

They now share `truncateForPreview(text, max = 200)` in the zero-import leaf `src/utils/format-utils.ts` (the issue's proposed natural home — it already holds `formatFileSize` and the timestamp formatters).

## Changes

- Add `truncateForPreview()` to `src/utils/format-utils.ts`, with a doc comment pinning the three encoded decisions.
- Replace the four call sites; `skill-tools.ts` keeps passing `params.description`.
- `test/utils/format-utils.test.ts`: boundary coverage (under / exactly at / over the limit, custom max, empty string) plus an equivalence test asserting byte-identical output against the old hand-rolled expression it replaces.
- Visible output is byte-identical — switching `...` to `…` or changing the limit is out of scope per the issue.
- Stacked on #1406: the `append_content` site moved to `src/tools/vault/append-content-tool.ts` in that PR, so one of the four sites appears at its new path.
- `src/services/vault-analyzer.ts:417`'s `substring(0, 200)` is left alone — a debug-log snippet with no ellipsis, explicitly out of scope.

## Screenshots / Screencast

Confirmation dialog output is byte-identical; nothing visual changes.

### Required

- [x] I have read and agree to the [Contributing Guidelines](../CONTRIBUTING.md)
- [x] I have read and agree to the [AI Policy](../AI_POLICY.md)
- [x] This PR is linked to an approved issue where the approach was discussed with a maintainer
- [x] All CI checks pass — `npm test` 3797 pass, `npm run build`, `npm run format-check`, `npm run lint:cycles`, `npm run typecheck:test`, `npm run knip`
- [ ] I have tested this change on Desktop
- [ ] I have verified this change does not break Mobile (or includes appropriate platform guards)
- [x] Documentation has been updated (if applicable) — no user-visible change; output byte-identical by design

### AI-Generated Code

- [x] This PR includes AI-generated or AI-assisted code
- [x] AI tool(s) used: pi coding agent (GLM)
- [x] I have reviewed and understand all AI-generated code in this PR

## Documentation updated

- None needed — no user-visible behavior change (byte-identical output).

## Summary by CodeRabbit

* **New Features**
  * Added tools for updating Markdown frontmatter properties and appending or replacing file content.
  * Vault editing tools now support clearer confirmations, progress messages, diffs, and error handling.

* **Improvements**
  * Standardized previews for memory, skill, file, and vault changes, truncating long content consistently.
  * Consolidated vault editing tools for more predictable availability and registration.

* **Tests**
  * Expanded coverage for vault editing, content previews, file handling, confirmations, and error scenarios.

### CODE REVIEW

File: src/tools/vault/update-frontmatter-tool.ts [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+import { t } from '../../i18n';
+import { getRawErrorMessageOr } from '../../utils/error-utils';
+
+/**
+ * Tool to safely update YAML frontmatter without touching content
+ * Critical for integration with Obsidian Bases and other metadata-driven plugins
+ */
+export class UpdateFrontmatterTool implements Tool {
+	name = 'update_frontmatter';
+	displayName = 'Update Frontmatter';
```
[the remark this answers is not in the document]
Comment (allenhutchison) [reply src/tools/vault/update-frontmatter-tool.ts:14]: Declining on this PR — the flagged file/strings come from the underlying #1310 move commits that this branch is stacked on, not from this PR's change (the preview-truncation dedup). The tool-layer i18n gap across every tool (all 11 vault tools share the same English `displayName`/`getProgressDescription` convention) is already tracked as #1391 and will be swept together there; localizing the two moved tools piecemeal here would leave the package inconsistent.

**Comment by allenhutchison** [comment]
Resolved the merge conflict with master after #1406 squash-merged (9166a78) — the add/add artifact on `append-content-tool.ts` resolved by keeping this PR's `truncateForPreview` import + preview change on top of the merged file. PR diff now contains only the #1292 change; build + full test suite (3803 tests), format, cycles, and typecheck all green on the new head.
